pkg/controller/machineset.go

```diff
@@ -697,37 +704,54 @@ func (c *controller) terminateMachines(ctx context.Context, inactiveMachines []*
 	Manipulate Finalizers
 */
 
-func (c *controller) addMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet) {
+func (c *controller) addMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet) error {
 	clone := machineSet.DeepCopy()
 
 	if finalizers := sets.NewString(clone.Finalizers...); !finalizers.Has(DeleteFinalizerName) {
 		finalizers.Insert(DeleteFinalizerName)
-		c.updateMachineSetFinalizers(ctx, clone, finalizers.List())
+		if err := c.updateMachineSetFinalizers(ctx, clone, finalizers.List()); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
-func (c *controller) deleteMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet) {
+func (c *controller) deleteMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet) error {
 	clone := machineSet.DeepCopy()
 
 	if finalizers := sets.NewString(clone.Finalizers...); finalizers.Has(DeleteFinalizerName) {
 		finalizers.Delete(DeleteFinalizerName)
-		c.updateMachineSetFinalizers(ctx, clone, finalizers.List())
+		if err := c.updateMachineSetFinalizers(ctx, clone, finalizers.List()); err != nil {
+			return err
+		}
 	}
+	return nil
 }
 
-func (c *controller) updateMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet, finalizers []string) {
-	// Get the latest version of the machineSet so that we can avoid conflicts
-	machineSet, err := c.controlMachineClient.MachineSets(machineSet.Namespace).Get(ctx, machineSet.Name, metav1.GetOptions{})
-	if err != nil {
-		return
-	}
+// updateMachineSetFinalizers tries to update the machineSet finalizers for a particular number of times
+func (c *controller) updateMachineSetFinalizers(ctx context.Context, machineSet *v1alpha1.MachineSet, finalizers []string) error {
+	var err error
+	machineSetName := machineSet.Name
```

Comment:
Isn't this variable redundant?

Reply:
yes initially I didn't have it , and used `machineSet.Name`, but on running against cluster , it gave `nil pointer Deref` error on line 755, so I had to use this variable instead.

Reply:
From the logic, either the loop should end successfully or there shall be a return with error. Either way, `machineSet` object should have retained its original value. Isn't it?

I have two solution proposals here - 

1. Can you check it in the Unit Tests? May be, by calling `updateMachineSetFinalizers` method with a `testMachineSet` having a wrong `namespace`, may be. 
2. In L739, make it as `machineSet, err := c.controlMachineClient.MachineSets(machineSet.Namespace).Get(ctx, machineSet.Name, metav1.GetOptions{})` -, you know, use the walrus operator for assignment & this should help retain the respective scopes of the two `machineSets` objects in the method.

Reply:
I re-looked at it. It happened that during testing I had commented some code and `err` was nil , and so on printing `null Deref` error was because of it.

Reply:
Good that you found it. Thanks!